[PeanutButters93] ip

PeanutButters93 opened the pull request:
# Duke2
> electric boogaloo

Has the following features
- Storing tasks
- Marking them as finished/unfinished
- Attaching dates to them

~test~

With multiple task types
1. Todos
2. Deadlines
3. Events

With CLI such as:
```
todo eat bread
list
bye
````


And they looks **somewhat** like this 👍 
- [ ] 1. eat [bread](https://en.wikipedia.org/wiki/Bread)
- [ ] 2. do homework by 2023-02-01 (Wed)

Commit:
Add Gradle support

PeanutButters93 committed:
Changed Duke.java. Added Greet, Echo and Exit functionality.
Updated gitignore include class files in src/main/java

PeanutButters93 committed:
Changed built-in phrases to give Duke a formal personality.
Edited the chatbox to have an indentation and max 60 character limit.

PeanutButters93 committed:
Added Add and List functionality

PeanutButters93 committed:
Added a Task class and added Marking as done functionality

PeanutButters93 committed:
Refactored code to use Regex and if statements instead of switch case and if statements. Changed some variables names. Fixed some syntax.

PeanutButters93 committed:
Added 3 subclasses of Tasks

PeanutButters93 committed:
Added test cases. Fixed the regex to search only from start of line.

PeanutButters93 committed:
Added some error handling. Automated Text UI broke in the process

PeanutButters93 committed:
Automated UI Testing fixed

PeanutButters93 committed:
Added Delete function

PeanutButters93 committed:
Update Duke.java to have a log between uses

PeanutButters93 committed:
Update Deadline and Event to use LocalDate

PeanutButters93 committed:
Merge branch 'branch-level-7'

PeanutButters93 committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	.gitignore
#	src/main/java/Deadline.java

PeanutButters93 committed:
Updated Deadline.java merging errors

PeanutButters93 committed:
Refactor entire code in OOP and gradle

PeanutButters93 committed:
Add package for all java files

PeanutButters93 committed:
Create Task Package

PeanutButters93 committed:
Edit parser. Mild changes to remove warnings

PeanutButters93 committed:
Revert "Updated Deadline.java merging errors"

This reverts commit 12f5da3d8903978a1db7e6909fc6d87022c45e3a.

PeanutButters93 committed:
Revert "Revert "Updated Deadline.java merging errors""

This reverts commit 33d8fc84334f44b90f28f39819a841d88c10d86a.

# Conflicts:
#	src/main/java/Deadline.java

PeanutButters93 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

PeanutButters93 committed:
Move date parser from Task to Parser
Add ability to export to jar w/ Gradle

PeanutButters93 committed:
Add tests for deadline

PeanutButters93 committed:
Add Find command

salty-flower commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,90 @@
+
+        switch (argFrags[0]) {
+            case "bye":
+                ui.print("Goodbye. Shutting down.");
```
Perhaps use named constant instead of a magic literal here (and many other places)?

PeanutButters93 replied:
Thanks for the advice, I've never heard of this term or that it was that bad a programming practice!

salty-flower commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.regex.Pattern;
+
+public class Deadline extends Task {
+    public static final Pattern pattern = Pattern.compile("(.+) /by (.+)");
```
Maybe name the constant in _uppercase using underscore_ style (i.e. `PATTERN`)?

PeanutButters93 replied:
Thanks for the catch! I forgot about renaming constants to be in caps

salty-flower commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,79 @@
+
+        Matcher m;
+        switch (type) {
+            case 'T':
```
Maybe no indentation for case clauses here?

PeanutButters93 committed:
Add Javadoc

PeanutButters93 committed:
Update code to follow Coding Standard

PeanutButters93 committed:
Merge branch 'branch-A-JavaDoc'

PeanutButters93 committed:
Merge branch 'branch-A-CodingStandard'

PeanutButters93 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java

PeanutButters93 committed:
Update Parser.java to fix find invalid format bug

PeanutButters93 committed:
Edit task subclasses to have protected fields

PeanutButters93 committed:
Reduced calls of System.out outside of Ui to prepare for GUI

PeanutButters93 committed:
Refactor code for GUI

jytheawesome commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+import java.time.format.DateTimeParseException;
+
+public class Duke {
+    private final Parser parser;
+    private final TaskList tasks;
```
Great job for making your fields private!

jytheawesome commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,111 @@
+        } catch (InvalidFormatException exception) {
+            return exception.getMessage();
+        } catch (UnrecognisedCommandException exception) {
+            return "Command not recognised. Please try again.";
```
Should such actions be handled by a `Ui` class? Not sure if you have since implemented it. Please do correct me if I am wrong.

jytheawesome commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,56 @@
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.LinkedList;
+import java.util.Scanner;
```
Would it be good to separate imports from different packages by a blank line?

jytheawesome commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+     */
+    public String find(String keyword) {
+        int i = 0;
+
```
Would it be better to not have a blank line here? When do you decide when to use blank lines?

jytheawesome commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,50 @@
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.HBox;
```
Is there a need to better group the imports here? Correct me if i'm wrong.

jytheawesome commented on src/main/java/duke/gui/DialogBox.java:
```diff
@@ -0,0 +1,50 @@
+    private Label text;
+    private ImageView displayPicture;
+
+    public DialogBox(Label l, ImageView iv) {
+        text = l;
+        displayPicture = iv;
+
+        text.setWrapText(true);
+        displayPicture.setFitWidth(100.0);
+        displayPicture.setFitHeight(100.0);
+
+        this.setAlignment(Pos.TOP_RIGHT);
+        this.getChildren().addAll(text, displayPicture);
+    }
+    private void flip() {
+        this.setAlignment(Pos.TOP_LEFT);
+        ObservableList<Node> tmp = FXCollections.observableArrayList(this.getChildren());
+        FXCollections.reverse(tmp);
+        this.getChildren().setAll(tmp);
+    }
+
+    public static DialogBox getUserDialog(Label l, ImageView iv) {
+        return new DialogBox(l, iv);
+    }
+
+    public static DialogBox getDukeDialog(Label l, ImageView iv) {
+        var db = new DialogBox(l, iv);
+        db.flip();
+        return db;
+    }
+}
```
good use of naming conventions for your methods and variables!

jytheawesome reviewed (commented):
Apologies for the late pull request review. Hope that my comments can be of relevance.

Overall good naming conventions, use of OOP, access modifiers, and packaging. . Would it be better to have certain actions like returning Strings of error messages or printing messages be handled by some sort of `Ui` or `Message` class? Feel free to let me know your thoughts on this. 

Keep up the good work!

PeanutButters93 committed:
Implement FXML

PeanutButters93 committed:
Fix UI warnings

PeanutButters93 committed:
Refactor duke package

Most of the code had poor SLAP practice. Could make it hard to read

Having only 1 layer of abstraction makes understanding functions
easier.

Let's extract complicated functions into separate parts.

PeanutButters93 committed:
TaskList class: Fix reading from log file bug

The TaskList writes to the log file in a simple, easy to read way.
But it reads it as if it was a normal input

The intended functionality is not working

Let's change to code to work temporarily.

PeanutButters93 committed:
Storage class : Extract updateLogFile to TaskList

Storage class uses Task package only for updateLogFile. Task List
uses it for a lot of functions.

Extracting updateLogFile has the benefit of making Storage simpler
while not affecting TaskList much.

Let's extract updateLogFile to TaskList

PeanutButters93 committed:
Storage Class: rename getIoException to getIOException

Originally Io to follow camel case, changed to IO to match original
class name

PeanutButters93 committed:
duke package: Add javadocs
Javadocs is not done for all methods in duke package

Let's add descriptions for them

PeanutButters93 committed:
Delete unused methods

PeanutButters93 committed:
Task Class : Fix error in javadoc

PeanutButters93 committed:
TaskList Class: Change loadFromStorage() to use stream
Currently uses a for loop

Slow and can be improved.

Let's change it to streams

PeanutButters93 committed:
Add assertions
Some segments of code rely on others

Another person taking over might not know

Let's use asserts to inform future people taking over the code

PeanutButters93 committed:
Merge branch 'branch-A-Assertions'

PeanutButters93 committed:
Merge branch 'branch-A-Streams'

PeanutButters93 committed:
Merge branch 'branch-A-CodeQuality'

PeanutButters93 committed:
Added New Text Document.txt
I don't have enough practice writing commits (Current State)

Writing commits in the recommended format will make further readers
understand the What, Why and How. (Benefit)

Let's create a dummy commit (Action)

PeanutButters93 committed:
Merge branch 'branch-Level-10'

PeanutButters93 committed:
Populated New Text Document.txt
Dummy commit

PeanutButters93 committed:
Populated New Text Document.txt with b

PeanutButters93 committed:
Merge pull request #2 from PeanutButters93/dummy-branch-a

Populated New Text Document.txt

PeanutButters93 committed:
Merge branch 'master' into dummy-branch-b

PeanutButters93 committed:
Merge pull request #3 from PeanutButters93/dummy-branch-b

Populated New Text Document.txt with b

PeanutButters93 committed:
TaskList class : Add regex to find for advanced users
Find currently uses string's contain to check for tasks

It can be improved to search for similar strings - such as brick or
brock

Let's implement regex into the find keyword

PeanutButters93 committed:
Remove New Text Document.txt
We still have this file

It was for practicing PR and commits. Serves no purpose now

Let's delete it

PeanutButters93 committed:
Update GUI to look better when resized
Elements do not auto resize when window size is changed

Is unintuitive and ugly

Let's change the fxml files

PeanutButters93 committed:
Fix bug where Date was not in output date format

PeanutButters93 committed:
Fix bug with Deadline and Event classes
Regex pattern was not being used correctly.

Unable to create Events or create Deadlines that didn't end with /by

Let's edit the code to fix that

PeanutButters93 committed:
Remove select comments

PeanutButters93 committed:
Add Ui.png

PeanutButters93 committed:
Move README.md to docs folder

PeanutButters93 committed:
Update README
Website is not informative

Could provide users with information to make learning the app faster

Let's add a user manual.

PeanutButters93 committed:
Update README.md to show the correct image

PeanutButters93 committed:
Fix javaFx19 warning

PeanutButters93 committed:
Add ParserTest
